[f] Explanation system

catriuspham opened the pull request:
### Status
Done

### Related issues
https://github.com/EastAgile/robber.py/issues/39

### Description
Let's implement an explanation system, where all the failure messages will have the following format, from which the second and the third line can be omitted. 

```python
BadExpectation:
A = ...
B = ...
C = ...
Expected A to {action} B. 
{additional information}
```

ea-open-source committed:
[f] First implementation of explanation system

- Apply new explanation system to to.be.eq
- Remove custom failure message (need to create the equivalent for
explanation system later)

catriuspham commented:
@NhanHo @hieueastagile Please take a look

ea-open-source committed:
[r] Split TestExplanation into multiple test cases

ea-open-source committed:
[f] Apply new explanation system to boolean matcher

- add negative_action to Explanation (this is convenient for
True/False, Truthy/Falsy... matchers)

hieueastagile commented:
Looks good to me so far, but I want to see is the explanation for comparing 2 objects (do the diff like what `sure` does)

ea-open-source committed:
[f] Show the diffs of two strings when comparing

catriuspham commented:
@oyster Could you please somehow turn off the comments from `coveralls` every time one pushes his code? It spams too much and I think the `checks` table at the end of the PR has enough information.

oyster commented:
It should be configurable in Coveralls settings. Please check if you can do that, @catriuspham. Or you can ask @hieueastagile to use our open source account to do it if you don't have privilege. Thanks.

hieueastagile commented:
lol, it's quite useful to me :)). Anyway, let me see if we can turn off it tomorrow.

ea-open-source committed:
[f] Apply Explanation for callable

ea-open-source committed:
[f] Apply Explanation for Called

ea-open-source committed:
[f] Apply Explanation for Length/Empty

ea-open-source committed:
Merge branch 'master' into feature/explaination-system

ea-open-source committed:
[f] Apply Explanation for CalledOnce

ea-open-source committed:
[f] Apply Explanation for Types

- Now Explanation can show the line 'A = None' when a None is passed
as argument `a`.

ea-open-source committed:
[f] Apply Explanation to RespondTo

ea-open-source committed:
[r] Make test_respond_to compatible with python 2.6

ea-open-source committed:
[f] Apply Explanation to Above, Below, Within

ea-open-source committed:
[f] Change Instanceof/instanceof into InstanceOf/instance_of

This is more grammatically correct

ea-open-source committed:
[f] Apply Explanation to InstanceOf

ea-open-source committed:
[f] Apply Explanation to Identical

ea-open-source committed:
[f] Apply Explanation to Exception

ea-open-source committed:
[r] Make test_instance_of compatible with Python 2.6

- Rename test_instanceof.py into test_instance_of.py

ea-open-source committed:
[r] Apply Explanation to Contain/Exclude

ea-open-source committed:
[f] Apply Explanation to CalledWith

- Add forced_disabled_repr to Explanation

ea-open-source committed:
[f] Add need_to_build_diffs to Explanation

- Remove special_init from Explanation

ea-open-source committed:
[f] Apply Explanation to EverCalledWith

ea-open-source committed:
[f] Apply Explanation to CalledOnceWith

ea-open-source committed:
[r] Remove commented code

ea-open-source committed:
[f] Apply Explanation to Change

- Explanation supports param 'd'

ea-open-source committed:
[f] Better explanation for negative cases

- When a explanation is negative, and it needs additional information,
it will add 'But it happened' at the last sentence
- From now, z will be the additional param

ea-open-source committed:
[r] Add `Diffs:` at the beginning of the diffs string

ea-open-source committed:
[r] Remove getting failure message logic in Base

Now Base only gets explanation

ea-open-source committed:
[r] Rename test_failure_message to test_explanation_message

ea-open-source committed:
[r] Remove commented code

ea-open-source committed:
[r] Update docs, and chain matcher test with Explanation

ea-open-source committed:
[f] Add CustomExplanation

ea-open-source committed:
[f] Add docstrings for Explanation

nhanbt commented:
@catriuspham check check

catriuspham commented:
@hieueastagile @NhanHo @nhanbt I finished this PR, please take a look when you guys have time

hieueastagile commented on README.md:
````diff
@@ -67,12 +67,12 @@ Asserts that the target is False:
 ```
 
-#### instanceof
+#### instance_of
````
lol, why did you change this, guy?

catriuspham replied:
Pycharm keeps yelling that there is a grammatical mistake there. Moreover, we have `respond_to`, `called_with` so `instance_of` will be more suitable.

hieueastagile replied:
I don't think it's a good reason for changing this, a small change in your lib make people have a lot of time to fix this in their code. I think it's better to revert this and we can consider it later in some major release.

seri-ea replied:
Yeah @catriuspham definitely should revert this change. It's breaking backwards compatibility for no reason. 

`instanceof` is also commonly written that way. Like the `sizeof` operator. They are treated exceptionally.

catriuspham replied:
I see. Let me do it right away. Thanks guys :)

hieueastagile commented on robber/explanation.py:
```diff
@@ -0,0 +1,106 @@
+        :param need_to_build_diffs: trigger if you want to build the diffs string
+        :param z: additional object ex: Expected A to be called with B but actually called with Z 
+        """
+        self.a = a
```
I guess a is actual, b is expected and what is c? Can we have a better naming here?

catriuspham replied:
Sometimes, 2 objects are not enough. For example:
```python
expect(func).to.change(1).by(2)
# a = func
# b = 1
# c = 2
```

hieueastagile replied:
Thanks for the explanation, anyway, please fix the name, a, b, c is not a good way to name the variable.

seri-ea replied:
It definitely is not @catriuspham And if your method takes so many arguments, please consider passing optional arguments by a chaining method, very much like the prevailing coding style of this project, like so:

```python
class Explanation
    def __init__(self, actual, action):
        pass
  
    def expected(self, expected):
        self._expected = expected
        return self

    def build_diffs(self, flagged=True):
        self._build_diffs = flagged
        return self
```

And we use it for the Equal matcher like so:
```python
Explanation(actual, 'equal').expected(expected).build_diffs()
```

catriuspham replied:
@hieueastagile Renamed them to `actual`, `expected`, `another_expected`, `other`

catriuspham replied:
@seri-ea I wonder if `chaining` is pythonic. We chain in robber just to create an user-friendly interface.

hieueastagile replied:
You were right @catriuspham, chaining should not be overuse in python.

hieueastagile commented on robber/matchers/__init__.py:
```diff
@@ -5,12 +5,12 @@
+    'contain',
     'equal',
-    'ever_called'
+    'ever_called',
```
I didn't notice it before, but `called` is good enough, huh?

catriuspham replied:
No no, `called` and `ever_called` are 2 different things. `called` is equivalent to `mock.assert_called` while `ever_called` is equivalent to `mock.assert_any_call`

hieueastagile replied:
What are the differences between them?

catriuspham replied:
`assert_called(*args, **kwargs)`
Assert that the mock was called at least once.

`assert_any_call(*args, **kwargs)`
assert the mock has been called with the specified arguments.

The assert passes if the mock has ever been called, unlike `assert_called_with()` and `assert_called_once_with()` that only pass if the call is the most recent one, and in the case of `assert_called_once_with()` it must also be the only call.

Source: https://docs.python.org/3/library/unittest.mock.html

hieueastagile replied:
so I think that it's should be:
```python
expect(callable).to.have.any_call(...)
```
should work? Would it be easier (since it's similar to the api of mock) to reason?

catriuspham replied:
I thought of it but it just sounds weird and awkward. So I think, it's better to keep `ever_called_with` and add `have.any_call` as an equivalent chain.

hieueastagile replied:
We should keep 1 only, two obvious way to solve a problem is not the pythonic way.

catriuspham replied:
It's not like we have another class implemented, I just have to write 1-line code to register an equivalent chain. This case is the same as `greater_than`,  `above`, `more_than`. They're just there for user's convenience.

hieueastagile replied:
I know it's quite simple, but I prefer to follow the Python zen in this case. What I'm thinking that we should do the same thing to the other apis, to keep them pythonic, 1 obvious way for every case.

catriuspham replied:
@hieueastagile I will create a separated PR for this in order to avoid making this PR god.

ea-open-source committed:
[r] Swap the position of c and additional action
to make it more natural

catriuspham commented:
@seri-ea would you like to take a look? :stuck_out_tongue_closed_eyes:

ea-open-source committed:
[r] Rename params of Expectation to make it easier to understand

seri-ea commented:
I will @catriuspham

seri-ea commented on README.md:
```diff
@@ -320,18 +320,19 @@ expect(1 + 1).to.be within(1, 3)
+    
+    @property
+    def explanation(self):
+        return Explanation(self.actual, self.is_negative, 'equal', self.expected)
```
It looks like this will return an `Explanation` that has nothing to do with the `Chain` matcher in the example code.

catriuspham replied:
It does, you can take a look at [robber/matchers/base.py](https://github.com/EastAgile/robber.py/blob/master/robber/matchers/base.py) line 29:
```python
message = self.message or self.explanation.message
```

seri-ea replied:
@catriuspham Please check again. I'm pretty sure what you meant is like this:
```python
return Explanation(self.actual, self.is_negative, 'have chain', self.expected)
```

catriuspham replied:
Sorry, my bad. I misunderstood.

seri-ea commented on robber/custom_explanation.py:
```diff
@@ -1,7 +1,7 @@
-class failure_message():
+class CustomExplanation:
     """
     with failure_message('Custom failure message'):
```
`with CustomExplanation` here?

catriuspham replied:
Nice catch :+1:

seri-ea commented on robber/explanation.py:
```diff
@@ -0,0 +1,107 @@
+
+    @staticmethod
+    def build_diff(a, b):
+        if type(a) is not str or type(b) is not str:
```
We should definitely try to build diffs of arrays and dictionaries as well.

catriuspham replied:
Done

seri-ea commented on robber/matchers/truthy.py:
```diff
@@ -10,15 +11,9 @@ class Truthy(Base):
-        return message.format(actual=self.actual, adj=adj)
+    @property
+    def explanation(self):
+        return Explanation(self.actual, self.is_negative, action="be truthy", negative_action='be falsy')
```
Let's always use single quotes whenever possible since that seems to be the style of the original robber.py.

catriuspham replied:
TBH, there're a lot of wrongly used double quotes like this around :scream: Silly me :disappointed:

hieueastagile replied:
Whatever you are using, stick to 1, `'` and `"` are both good, but it needs to be consistent.

seri-ea commented:
@catriuspham I made some minor suggestions, but as a potential user of EA-robber.py, I find the explanation format a bit unfriendly. 

I mean, why:
```
BadExpectation:
A = 123
B = "123"
Expected A to equal B. 
````

Instead of this:
```
Expected <int>123 to equal <string>123
```

And yet it doesn't support what @hieueastagile was looking for (explanation for comparing 2 objects). Sure we can and should do that in another PR as a detailed implementation of the `build_diffs` method, but for now I wonder if we could take a step back and re-think the message format.

ea-open-source committed:
[f] Better way to deal with more detail

ea-open-source committed:
[r] Revert "Change Instanceof/instanceof into InstanceOf/instance_of"

ea-open-source committed:
[r] Replace wrongly used double quotes with single quotes

catriuspham commented:
@seri-ea the reason for the messages to have the format
```python
BadExpectation:
A = 123
B = "123"
Expected A to equal B. 
```
is that sometimes the represented string of A and B can be super long, so we need to find a consistent way to make them look as clear as possible.

For example, this message

```python
Expected <string>"Some super long texts that make your eyes bleed a lot until you faint" to equal <string>"Some super long texts that make your eyeballs bleed a lot until you faint"
```

look awkward.

But it's a lot better with the new format:

```python
BadExpectation:
A = Some super long texts that make your eyes bleed a lot until you faint
B = Some super long texts that make your eyeballs bleed a lot until you faint
Expected A to equal B. 
```

seri-ea commented:
@catriuspham Thanks for the explanation. The counter argument is that while your format makes it clear for when the string representations are long, it makes most cases harder and longer to parse (the reader has to look up A, B, C... kind of awkward), and that's probably why no popular testing library is using your failure message format yet.

Maybe only break down into lines for special cases such as the equality matcher, where you are going to have to do deep comparison and handle it exceptionally anyway.

ea-open-source committed:
[r] Small edit in Readme

ea-open-source committed:
[f] Build diff for list and dict

ea-open-source committed:
[r] Make the test run on python 3.5

catriuspham commented:
@seri-ea Actually, the idea of this format is taken from how `unittest` display its failure message, it's not something that I make up myself though. I agree with you that in some cases when the represented strings of the objects are short, the message looks awkward. But on the larger scale, to me, this is the best format I can think of to cover all the cases. To detect when to use the new format, when to keep the old one would require a lot of effort though.

hieueastagile commented:
@catriuspham Actually, you don't need a format which works for all cases, most of the messages are specific to its own assertion, we don't need to make it generic like you did. Anyway, I feel what you did are already good enough, not the best but usable, just let me know when you finish the diff for object/array, then we can merge it. I really want it to be merged soon and we will deploy it to pypi as robber 1.2.0. We should do another refactor later for robber 2.0, where we can think more carefully about the architecture.

catriuspham commented:
@hieueastagile Please give me 1-2 day to with the `none` param issue. Gonna notice you as soon as I finish it.

hieueastagile commented:
Take your time, no rush then @catriuspham

ea-open-source committed:
[f] None params are handle correctly

ea-open-source committed:
[f] Ignore .tox folder from flake8

catriuspham commented:
@hieueastagile All things are done. However, I guess we have to wait for `coverage`, it's been 4 hours though...

catriuspham commented:
@hieueastagile Everything's good. Please take a look and merge this PR whenever you want.

hieueastagile commented:
@catriuspham Let me test it a bit first, I will tell anh Quang to merge when I finished.
@NhanHo @nhanb @steadykhoi @hotay please help take a look on this PR also.

hieueastagile commented:
@catriuspham The diff here is still unclear to me, here is what I got when I tried to compare two unequal dictionary.
```python
Traceback (most recent call last):
  File "test-robber.py", line 8, in <module>
    'a': 1
  File "/home/hieu/.pyenv/versions/2.7.12/envs/robber.py/src/robber/robber/expect.py", line 44, in method
    return klass(self.obj, other, negative_fact, *args, **kwargs).fail_with(self.message).match()
  File "/home/hieu/.pyenv/versions/2.7.12/envs/robber.py/src/robber/robber/matchers/base.py", line 30, in match
    raise BadExpectation(message)
robber.bad_expectation.BadExpectation: 
A = {'a': 1, 'b': 2}
B = {'a': 1}
Expected A to equal B
Diffs:
- {'a': 1, 'b': 2}
+ {'a': 1}
```

I think it should be more clear, something like:
```python
A = {'a': 1, 'b': 2}
B = {'a': 1}
Expected A to equal B
Diffs:
- {'a': 1, 'b': 2}
+ {'a': 1}
A has 'b' key but B don't
```
if they had same key, it should be:
```
A['a'] is 1 but B['b'] is 2
```
We need to that then the developer who use the library can know where to start fixing their test. Wdyt?

hieueastagile commented:
Btw, can we somehow add the color in and remove the `traceback`?

catriuspham commented:
@hieueastagile Great ideas, let's see what I can do

catriuspham commented:
@hieueastagile @seri-ea @nhanbt 

> If you don't know what I'm talking about, please take a look at 3rd comment upward.

About the dict difference message. I'm thinking of 2 outputs. 
One is that you only notice the user about the very first difference this is what `Sure` is doing, for example:
```python
d1 = {'a': 1, 'b': 2, 'c': 3, 'd': 4}
d2 = {'a': 1, 'd': 2, 'e': 3, 'f': 4}
expect(d1).to.eq(d2)

Output: A has the key 'b' while B does not
```
Another case is that you list out all the differences, for the above example, the output would be:
```python
Output: 
A has the key(s) 'b', 'c' while B does not
A['d'] = 4 while B['d']=2
```
Both are fine for me, what's your preference?

hieueastagile commented:
@catriuspham Both are fine. The idea that we want to give the user a clear enough message, then they can start to fix their test, so I think both, let's stick with the simplest option and we can extend later if needed.

seri-ea commented:
Agree, both are fine, so whichever gives you an easy time to implement 😄

ea-open-source committed:
[f] Build diffs when comparing 2 dicts

ea-open-source committed:
[f] Move build diffs logic to Equal

Only Equal needs to build diffs, so let it do that by itself.

ea-open-source committed:
[f] Build diffs when comparing 2 lists

ea-open-source committed:
[r] Update some docstrings

catriuspham commented:
@hieueastagile I cannot find a way to properly remove the traceback. With some lines of code, I can make it disappear on the default python shell, but ipython is another story as it has its own way to discover traceback, the same lines of code make ipython yell out some error messages, and it's really annoying. For the color, I can manually colorize each part of the message, however, we need further discussion on what color to use, which part to be colorized... Wdyt?

hieueastagile commented:
@catriuspham : How about catching the exception and just print the message? I guess there's no traceback if we caught the exception? About color part, please take a look on sure and rspec of Rails. You can search for Youtube video to pick the color. In general, I think the rule is simple, green for whatever which passed and red for failing tests? Do we need any other colors?

catriuspham commented:
"catching the exception and just print the message" means that we will have to get rid of `BadExpectation`, as there are a lot of tests involved, this is not a small task though. Moreover, our current logic checks if a chain is "matched" and `raise` exceptions directly, so there's no way to properly `catch` the exception here without refactoring the `base` class, which is another big chore.
About the color, correct me if I'm wrong, but we currently print message "only" if some tests fail, which means every message will be in red. Also, I don't think `sure` colorize its messages.

hieueastagile commented:
IIRC, `sure` had that. It could be your configuration for not seeing it, anyway, see this https://www.youtube.com/watch?v=gsgG-JvXsJ0 (8:14) to see how RSpec did that.

Ok, so let's ignore the traceback, just create a story and we can consider it later. But I wonder if `sure` print the traceback?

catriuspham commented:
Sure does keeps the traceback.

ea-open-source committed:
[f] Colorize failure message (red)

catriuspham commented:
@hieueastagile I added color to the message, please check if it's what you meant.

catriuspham commented:
@hieueastagile About the color, I think making the message red is enough for now. Learning basics of Ruby and rspec needs time. I will raise an issue with more research and suggestions later. Also, this PR is quite big and lives for a long time (more than a month to be precise), keeping it around will decrease the work efficiency. Therefore, in my opinion, we'd better merge it and continue on other features.

hieueastagile commented:
@catriuspham Ok, I think it's time to merge this. You can ask Huy to help on rspec if need to.
@oyster Can you help us merge this branch?
